### PR TITLE
Make `asyncio.sleep(0)` run faster (take 2)

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

Trying again #4568 with a different approach. Mostly adapted from scheduler in [core-js](https://github.com/zloirock/core-js/blob/master/packages/core-js/internals/task.js) and [setimmediate.js](https://github.com/YuzuJS/setImmediate)

1. If Node, use `setImmidiate`. This API is deprecated in most browsers but exists in Node.
2. If in the browser main thread. Use `postMessage` + event listener
3. If in the web worker (except for safari), use `MessageChannel`.
4. Otherwise, fallback to `settimeout(0)`.

Unfortunetly, safari + web worker will fallback to `settimeout(0)` which is slow, but otherwise it works much faster.

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests

### CODE REVIEW

File: src/js/scheduler.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  const onGlobalMessage = (event: MessageEvent) => {
+    if (
+      typeof event.data === "string" &&
+      event.data.indexOf(scheduleCallbackImmediateMessagePrefix) === 0
+    ) {
+      const handle = +event.data.slice(
+        scheduleCallbackImmediateMessagePrefix.length,
+      );
+      const task = tasks[handle];
+      if (task) {
```
Comment (hoodmane) [inline src/js/scheduler.ts:30]: nit:
```js
if (!task) {
   return;
}
```

File: src/js/environments.ts [hunk 2]
```diff
@@ -31,10 +31,35 @@ export const IN_BROWSER_MAIN_THREAD =
...
+export function detectEnvironment(): Record<string, boolean> {
+  return {
+    IN_NODE: IN_NODE,
+    IN_NODE_COMMONJS: IN_NODE_COMMONJS,
+    IN_NODE_ESM: IN_NODE_ESM,
+    IN_DENO: IN_DENO,
+    IN_BROWSER: IN_BROWSER,
+    IN_BROWSER_MAIN_THREAD: IN_BROWSER_MAIN_THREAD,
+    IN_BROWSER_WEB_WORKER: IN_BROWSER_WEB_WORKER,
+    IN_SAFARI: IN_SAFARI,
```
Comment (hoodmane) [inline src/js/environments.ts:63]:
```suggestion
    IN_NODE,
    IN_NODE_COMMONJS,
    IN_NODE_ESM,
    IN_DENO,
    IN_BROWSER,
    IN_BROWSER_MAIN_THREAD,
    IN_BROWSER_WEB_WORKER,
    IN_SAFARI,
```

**Review by hoodmane — APPROVED** [review]
Thanks for working this out @ryanking13. It looks quite tricky.

**Comment by hoodmane** [comment]
See also:
https://github.com/WICG/scheduling-apis/blob/main/explainers/yield-and-continuation.md
https://developer.chrome.com/blog/introducing-scheduler-yield-origin-trial/

**Comment by hoodmane** [comment]
Perhaps we should check for `scheduler.yield()` and use that if it exists.

File: src/js/scheduler.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    tasks[nextTaskHandle] = callback;
+    globalThis.postMessage(
+      scheduleCallbackImmediateMessagePrefix + nextTaskHandle,
+      "*",
+    );
+    nextTaskHandle++;
+  } else if (IN_BROWSER_WEB_WORKER && !IN_SAFARI) {
+    const channel = new MessageChannel();
+    channel.port1.onmessage = () => callback();
+    channel.port2.postMessage("");
```
Comment (hoodmane) [inline src/js/scheduler.ts:74]: In the core-js implementation they only make one MessageChannel and share it for all messages. Is it really a good idea to allocate a separate MessageChannel every time we call this?

File: src/js/environments.ts [hunk 4]
```diff
@@ -31,10 +31,35 @@ export const IN_BROWSER_MAIN_THREAD =
...
   typeof document !== "undefined" &&
   typeof document.createElement !== "undefined" &&
-  typeof sessionStorage !== "undefined";
+  typeof sessionStorage !== "undefined" &&
+  typeof importScripts === "undefined";
 
 /** @private */
 export const IN_BROWSER_WEB_WORKER =
   IN_BROWSER &&
   typeof importScripts !== "undefined" &&
```
Comment (hoodmane) [inline src/js/environments.ts:40]: Can we change this to `typeof importScripts === "function"`

File: src/js/scheduler.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    // node has setImmediate, let's use it
+    setImmediate(callback);
+  } else if (IN_BROWSER_MAIN_THREAD) {
+    tasks[nextTaskHandle] = callback;
+    globalThis.postMessage(
+      scheduleCallbackImmediateMessagePrefix + nextTaskHandle,
+      "*",
+    );
+    nextTaskHandle++;
+  } else if (IN_BROWSER_WEB_WORKER && !IN_SAFARI) {
```
Comment (hoodmane) [inline src/js/scheduler.ts:71]: Can we also check that `typeof MessageChannel === "function"` here?

File: src/js/scheduler.ts (hunk 3 above)
Comment (ryanking13) [reply src/js/scheduler.ts:74]: I tried that in [this commit](https://github.com/pyodide/pyodide/pull/4590/commits/b72cf110b8d2fc76d1b59ce60b0c51ade1a3609a), but somehow there was a failure in Node.js. I don't understand why.

File: src/tests/test_webloop.py [hunk 6]
```diff
@@ -497,3 +497,62 @@ async def temp():
         loop._no_in_progress_handler = None
         loop._keyboard_interrupt_handler = None
         loop._system_exit_handler = None
+
+
+@run_in_pyodide
+async def test_zero_timeout(selenium):
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply src/tests/test_webloop.py:503]: No, safari + main thread seems to work fine with postMessage approach.

**Comment by ryanking13** [comment]
> Perhaps we should check for scheduler.yield() and use that if it exists.

Thanks for the info! I quickly checked if this is available in my local chrome, but it seems like `scheduler.yield` is under the experimental flag. While it's fine to add experimental flag for testing (and we already do for JSPI), let's revisit this when the feature becomes stable, as we have other options for now.

**Pull request merged**